vhdx/src/vhdx.rs

```diff
@@ -0,0 +1,228 @@
+        ) {
+            Ok(r) => (r.bat_entry, r.mdr_entry, r.region_entries),
+            Err(e) => {
+                return Err(Error::ParseVhdxRegionEntry(e));
+            }
+        };
+
+        let bat_entry = bat_entry.unwrap();
+        let mdr_entry = mdr_entry.unwrap();
+
+        let disk_spec = MetadataTableEntry::parse_metadata(&mut file, mdr_entry.clone())
+            .map_err(Error::ParseVhdxMetadata)?;
+        let bat_entries =
+            BatEntry::collect_bat_entries(&mut file, disk_spec.clone(), bat_entry.clone())
+                .map_err(Error::ReadBatEntry)?;
+
+        Ok(Vhdx {
+            file,
+            vhdx_header,
+            region_entries,
+            bat_entry,
+            mdr_entry,
+            disk_spec,
+            bat_entries,
+            current_offset: 0,
+        })
+    }
+
+    pub fn virtual_disk_size(&self) -> u64 {
+        self.disk_spec.virtual_disk_size
+    }
+}
+
+/// Check if a file is a VHDx disk
+pub fn is_vhdx(f: &mut File) -> std::io::Result<bool> {
+    match FileTypeIdentifier::new(f) {
+        Ok(_) => Ok(true),
+        Err(_) => Ok(false),
+    }
+}
+
+impl Read for Vhdx {
+    /// Wrapper function to satisfy the Read trait implementation for VHDx disk. Converts the offest to secotor index and buffer length to sector count.
+    fn read(&mut self, buf: &mut [u8]) -> std::result::Result<usize, std::io::Error> {
+        let sector_count =
+            div_round_up!(buf.len() as u64, self.disk_spec.logical_sector_size as u64);
+        let sector_index = self.current_offset / self.disk_spec.logical_sector_size as u64;
+
+        match vhdx_io::read(
+            &mut self.file,
+            buf,
+            &self.disk_spec,
+            &self.bat_entries,
+            sector_index,
+            sector_count,
+        ) {
+            Ok(r) => Ok(r),
+            Err(e) => {
+                println!("VHDx read: {}", e);
+                Err(io::Error::last_os_error())
+            }
+        }
+    }
+}
+
+impl Write for Vhdx {
+    fn flush(&mut self) -> std::result::Result<(), std::io::Error> {
+        self.file.flush()
+    }
+
+    /// Wrapper function to satisfy the Write trait implementation for VHDx disk. Converts the offest to secotor index and buffer length to sector count.
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, std::io::Error> {
+        let sector_count =
+            div_round_up!(buf.len() as u64, self.disk_spec.logical_sector_size as u64);
+        let sector_index = self.current_offset / self.disk_spec.logical_sector_size as u64;
+
+        match vhdx_io::write(
+            &mut self.file,
+            buf,
+            &self.disk_spec,
+            self.bat_entry.file_offset,
+            &mut self.bat_entries,
+            sector_index,
+            sector_count,
+        ) {
+            Ok(r) => Ok(r),
+            Err(e) => {
+                println!("Vhdx write: {}", e);
+                Err(io::Error::last_os_error())
```

Comment:
You want to use `map_err()?` as well here in order to avoid losing the error you got from `vhdx_io::write()`.